libtenzir/builtins/operators/show.cpp

```diff
@@ -0,0 +1,264 @@
+    },
+  };
+}
+
+struct operator_args {
+  located<std::string> aspect;
+
+  friend auto inspect(auto& f, operator_args& x) -> bool {
+    return f.object(x)
+      .pretty_name("operator_args")
+      .fields(f.field("aspect", x.aspect));
+  }
+};
+
+class show_operator final : public crtp_operator<show_operator> {
+public:
+  show_operator() = default;
+
+  explicit show_operator(operator_args args) : args_{std::move(args)} {
+  }
+
+  auto operator()(operator_control_plane& ctrl) const
+    -> generator<table_slice> {
+    auto contains = [](const auto& plugins, const auto& name) -> bool {
+      auto f = [&](const auto& plugin) {
+        return plugin->name() == name;
+      };
+      return std::find_if(plugins.begin(), plugins.end(), f) != plugins.end();
+    };
+    if (args_.aspect.inner == "connectors") {
+      auto loaders = collect(plugins::get<loader_parser_plugin>());
+      auto savers = collect(plugins::get<saver_parser_plugin>());
+      auto connectors = std::set<std::string>{};
+      for (const auto* plugin : loaders)
+        connectors.insert(plugin->name());
+      for (const auto* plugin : savers)
+        connectors.insert(plugin->name());
+      auto builder = table_slice_builder{connector_type()};
+      for (const auto& connector : connectors) {
+        if (not(builder.add(connector)
+                && builder.add(contains(loaders, connector))
+                && builder.add(contains(savers, connector)))) {
+          diagnostic::error("failed to add connector")
+            .note("from `show {}`", args_.aspect.inner)
+            .emit(ctrl.diagnostics());
+          co_return;
+        }
+      }
+      co_yield builder.finish();
+    } else if (args_.aspect.inner == "formats") {
+      auto parsers = collect(plugins::get<parser_parser_plugin>());
+      auto printers = collect(plugins::get<printer_parser_plugin>());
+      auto formats = std::set<std::string>{};
+      for (const auto* plugin : parsers)
+        formats.insert(plugin->name());
+      for (const auto* plugin : printers)
+        formats.insert(plugin->name());
+      auto builder = table_slice_builder{format_type()};
+      for (const auto& format : formats) {
+        if (not(builder.add(format) && builder.add(contains(parsers, format))
+                && builder.add(contains(printers, format)))) {
+          diagnostic::error("failed to add format")
+            .note("from `show {}`", args_.aspect.inner)
+            .emit(ctrl.diagnostics());
+          co_return;
+        }
+      }
+      co_yield builder.finish();
+    } else if (args_.aspect.inner == "operators") {
+      auto builder = table_slice_builder{operator_type()};
+      for (const auto* plugin : plugins::get<operator_parser_plugin>()) {
+        // TODO: figure out how we can get the operator type. Ideally also it's
+        // arguments.
+        if (not(builder.add(plugin->name()) && builder.add(false)
+                && builder.add(false) && builder.add(false))) {
```

Comment:
@jachris is this information easily available?

Reply:
Not really. The possible input types can depend on the arguments (e.g., `remote ...`), so in general, you cannot decide this purely by using the operator plugin. For the arguments, we wanted to have them as data anyway, but this is currently not a priority and requires quite a bit of effort to get right.

Reply:
Just to be clear: with "operator type" I meant whether it's a source, sink, or transform. Not the input/output types.

Reply:
Ah, I see, but the same applies here (`remote` can be a source, sink, or transform, but we don't know in advance).

Reply:
So then all apply, no?

The point is to give the user more information on how to use operators. For many, it's only 1 or 2, so this is very helpful information to have.

Reply:
> So then all apply, no?

Depends on your definition of "source", etc. But sure, you could say so.

> What I am asking: can we make plugin API changes so that we can query this?

We can. But then we should make the plugins return the possible type signatures. When using the `crtp_operator`, we can even infer most of this automatically. How important is this to you?

Reply:
> How important is this to you?

This is probably a continuous effort. I would like to get something (marked as experimental) through the door quickly to enable our partners so that they can use these introspection capabilities, even if they are incomplete or have warts.

Anything that's scheduling-worth effort should be a strict follow-up. Within the scope of this PR, I'd just like to ship a something-is-better-than-nothing outcome.